pkg/storage/unified/resource/sqlkv.go

```diff
@@ -298,7 +330,185 @@ func (k *sqlKV) TestingSave(ctx context.Context, key string, value []byte) error
+	if w.sectionKey.Section == eventsSection {
+		_, err := dbutil.Exec(w.ctx, w.kv.db, sqlKVSaveEvent, sqlKVSaveRequest{
+			SQLTemplate:     sqltemplate.New(w.kv.dialect),
+			sqlKVSectionKey: w.sectionKey,
+			Value:           w.buf.Bytes(),
+		})
+
+		if err != nil {
+			return fmt.Errorf("failed to save: %w", err)
+		}
+
+		return nil
+	}
+
+	// if storage_backend is running with an RvManager, it will inject a transaction into the context
+	// used to keep backwards compatibility between sql-based kvstore and unified/sql/backend
+	tx, ok := rvmanager.TxFromCtx(w.ctx)
+	if !ok {
+		// temporary save for dataStore without rvmanager
+		// we can use the same template as the event one after we:
+		// - move PK from GUID to key_path
+		// - remove all unnecessary columns (or at least their NOT NULL constraints)
+		_, err := w.kv.Get(w.ctx, w.sectionKey.Section, w.sectionKey.Key)
+		if errors.Is(err, ErrNotFound) {
+			_, err := dbutil.Exec(w.ctx, w.kv.db, sqlKVInsertData, sqlKVSaveRequest{
+				SQLTemplate:     sqltemplate.New(w.kv.dialect),
+				sqlKVSectionKey: w.sectionKey,
+				GUID:            uuid.New().String(),
+				Value:           w.buf.Bytes(),
+			})
+
+			if err != nil {
+				return fmt.Errorf("failed to save: %w", err)
+			}
+
+			return nil
+		}
+
+		if err != nil {
+			return fmt.Errorf("failed to save: %w", err)
+		}
+
+		_, err = dbutil.Exec(w.ctx, w.kv.db, sqlKVUpdateData, sqlKVSaveRequest{
+			SQLTemplate:     sqltemplate.New(w.kv.dialect),
+			sqlKVSectionKey: w.sectionKey,
+			Value:           w.buf.Bytes(),
+		})
+
+		if err != nil {
+			return fmt.Errorf("failed to save: %w", err)
+		}
+
+		return nil
+	}
+
+	// special, temporary save that includes all the fields in resource_history that are not relevant for the kvstore,
+	// as well as the resource table. This is only called if an RvManager was passed to storage_backend, as that
+	// component will be responsible for populating the resource_version and key_path columns
+	// note that we are not touching resource_version table, neither the resource_version columns or the key_path column
+	// as the RvManager will be responsible for this
+	dataKey, err := ParseKeyWithGUID(w.sectionKey.Key)
+	if err != nil {
+		return fmt.Errorf("failed to parse key: %w", err)
+	}
+
+	var action int64
+	switch dataKey.Action {
+	case DataActionCreated:
+		action = 1
+	case DataActionUpdated:
+		action = 2
+	case DataActionDeleted:
+		action = 3
+	default:
+		return fmt.Errorf("failed to parse key: %w", err)
+	}
+
+	_, err = dbutil.Exec(w.ctx, tx, sqlKVInsertLegacyResourceHistory, sqlKVSaveRequest{
+		SQLTemplate: sqltemplate.New(w.kv.dialect),
+		Value:       w.buf.Bytes(),
+		GUID:        dataKey.GUID,
+		Group:       dataKey.Group,
+		Resource:    dataKey.Resource,
+		Namespace:   dataKey.Namespace,
+		Name:        dataKey.Name,
+		Action:      action,
+		Folder:      dataKey.Folder,
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to save: %w", err)
+	}
+
+	switch action {
```

Comment:
We could switch on `dataKey.Action` to make the `case`s more readable using the corresponding constants.